README
class-dump is a command-line utility for examining the Objective-C
segment of Mach-O files.  It generates declarations for the classes,
categories and protocols.  This is the same information provided by
using 'otool -ov', but presented as normal Objective-C declarations.

The latest version and information is available at:

    <http://www.codethecode.com/projects/class-dump>

-----

This file is part of class-dump, a utility for examining the
Objective-C segment of Mach-O files.
Copyright (C) 1997-1998, 2000-2001, 2004-2011 Steve Nygard.

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

You may contact the author by:
   e-mail:  class-dump at codethecode.com


CHANGELOG.md
### Unreleased

* Building with the 10.14 SDK.
* Fix parse error of _Atomic types.  (Thanks to Jacob Bandes-Storch (jtbandes)).
* Work around many other parsing errors from blank types:
  * types like vector_float4 have a blank encoding. (Thank you iTerm2!  Open source FTW.)
  * Swift types have a blank encoding.
  * Arrays of the above types failed because... no type
  * Now, instead of generating a parse error, the parser just assumes a type is missing and uses one called "MISSING_TYPE".
* Switched from OpenSSL to CommonCrypto for decryption.
* Replaced CommonCrypto Blowfish decryption (which checks for valid key sizes) with a C implementation that doesn't, since
  we need to use bigger keys.
* Some other changes from Eric Monti, Oleksiy Ivanov, and Cédric Luthi

### Version 3.5 - Released 2013-11-16

* Targeting 10.8+ now, and building with the 10.9 SDK.
* Fixed Objective-C 2.0 ivar offsets. It now displays actual offsets instead of ivar addresses.
* Support for the new `arm64` architecture.
  * Recognize the `--arch arm64` option
  * Display `arm64` instead of `0x100000c:0x0`
  * Handle the new `LC_ENCRYPTION_INFO_64` load command
* Parse extended types when available. It means that you can extract more precise type information on protocol methods. E.g. you get `- (NSArray *)bookmarksFromContentItems:(NSArray *)arg1;` instead of `- (id)bookmarksFromContentItems:(id)arg1;`. Unfortunately, extended types are not available on classes, only on protocols.
* Parse blocks type information, i.e. you get `CDUnknownBlockType` instead of `id`. Inside protocol methods – thanks to extended types – you get the full block signature, e.g. `- (void)accountsWithHandler:(void (^)(NSArray *, NSError *))arg1;`.
* Resolve `@loader_path`. Xcode and the developer tools use this.
* Added the `--hide` options to hide struct/union and protocol sections. Use `--hide structures` and `--hide protocols` or `--hide all` to hide both of those. This is meant for development, to make it easier to see just the other sections.
* Tweaked comments, now using `//` instead of `/* */` in generated headers.
* Show a better error message when trying to run class-dump on a static library.
* Fix the way the ObjC 2 list header `entsize` field is read. This allows to run class-dump on files produced by the latest version of [dyld_decache](https://github.com/kennytm/Miscellaneous#dyld_decache).
* Parse the `LC_FUNCTION_STARTS` load command.

### Version 3.4 - Released 2012-11-19

* Targeting 10.7+ now. Needs to build with the 10.8 SDK.
* Lots of changes to the code. Style changes. Using automatic reference counting. Objective-C literals and array/dictionary indexing.
* Recognise `armv7s` architecture.
* Fixed a bug where `--arch x86_64` wouldn’t find the correct architecture.
* Handle 4 new load commands from 10.8:
  * `LC_MAIN`
  * `LC_DATA_IN_CODE`
  * `LC_SOURCE_VERSION`
  * `LC_DYLIB_CODE_SIGN_DRS`
* Show source and SDK versions.
* Fix crasher on SDK frameworks, which have empty sections.
* deprotect fat files, endian neutral, better error messages. [0xced]
* Use `__attribute__((visibility(“hidden”)))` instead of `// Not exported` [0xced]
* Show a more informative error message when the file doesn’t contain the desired architecture.
* Switch to `NSRegularExpression` for `-C` arg matching.

### Version 3.3.4 - Released 2011-09-03

* Fixed assertion failure on many frameworks in Lion.
* Replaced SDK aliases with new command line options, `--sdk-ios` and `--sdk-mac`. Now you can use `--sdk-ios 5.0`
* Handle another type: @encode(long double) = D
* Handle `_Complex`. This finally fixes the parse error with Grapher.app
* Added support for minimum Mac OS X and iOS versions (`LC_VERSION_MIN_MACOSX`, `LC_VERSION_MIN_IPHONEOS`).
* Show `LC_DYLD_ENVIRONMENT` information.
* Updated deployment target to 10.6
* Fixed logs about some unknown load commands.
  * `LC_LOAD_UPWARD_DYLIB` - treat this just like `LC_LOAD_DYLIB`
  * `LC_FUNCTION_STARTS` - we don’t do anything with this command, just stopped the warning
* I’ve switched to git, and moved the repository to github. The latest source is now available at the [class-dump git repository at github](http://github.com/nygard/class-dump).

#### Known Bugs

* Assertion failure on an older version Kindle.app
  * This may be caused by optional methods in protocols.

### Version 3.3.3 - Released 2010-08-08

* Fixed problems where many iOS class names were missing or appeared as “(null)”/”__objc_empty_cache”.
* Added `--sdk-root` option to let you specify a root path to search for frameworks. You can use a path, or short aliases: 4.1, 4.0, 3.2, 10.6, 10.5
* Resolve and use run paths, and show each resolved path. Xcode and the developer tools use this.
* Report the correct name of architectures that use 64 bit ABI.
* A few visible changes contributed by Cédric Luthi, and many more in the code:
  * Added a comment above @interface of classes that are not exported.
  * Fix to find first read/write segment (i.e. Accelerate i386 framework).
  * Show UUID under filename.
* ~~The latest source is now available at the [class-dump Mercurial repository at bitbucket.org](http://bitbucket.org/nygard/class-dump).~~
* 2011-05-12: I’ve switched to git, and moved the repository to github. The latest source is now available at the [class-dump git repository at github](http://github.com/nygard/class-dump).
 
### Version 3.3.2 - Released 2010-05-11

* The `-a` and `-A` options should work again.
* Better error reporting for non-existent and unreadable files. (Patch from Gerriet M. Denkmann)
* Added missing classes to one of the targets.

### Version 3.3.1 - Released 2009-09-17

* Handle 10.6 style protected binaries, so that you can once again class-dump Finder, Dock, SystemUIServer, etc.
* Fixed crasher when trying to dump 64-bit files with 32-bit class-dump. This now prints an error message.
* Generate and use typedefs for template types that are used in methods.

### Version 3.3 - Released 2009-09-01


main.m
// -*- mode: ObjC -*-

//  This file is part of class-dump, a utility for examining the Objective-C segment of Mach-O files.
//  Copyright (C) 1997-1998, 2000-2001, 2004-2010 Steve Nygard.

#include <stdio.h>
#include <libc.h>
#include <unistd.h>
#include <getopt.h>
#include <stdlib.h>
#include <mach-o/arch.h>

#import <Foundation/Foundation.h>
#import "NSString-Extensions.h"

#import "CDClassDump.h"
#import "CDFindMethodVisitor.h"
#import "CDClassDumpVisitor.h"
#import "CDMultiFileVisitor.h"
#import "CDFile.h"
#import "CDMachOFile.h"
#import "CDFatFile.h"
#import "CDObjectiveC2Processor64.h"
#import "CDFatArch.h"
#import "CDSearchPathState.h"

void print_usage(void)
{
    fprintf(stderr,
            "class-dump %s\n"
            "Usage: class-dump [options] <mach-o-file>\n"
            "\n"
            "  where options are:\n"
            "        -a             show instance variable offsets\n"
            "        -A             show implementation addresses\n"
            "        --arch <arch>  choose a specific architecture from a universal binary (ppc, ppc64, i386, x86_64)\n"
            "        -C <regex>     only display classes matching regular expression\n"
            "        -f <str>       find string in method name\n"
            "        -H             generate header files in current directory, or directory specified with -o\n"
            "        -I             sort classes, categories, and protocols by inheritance (overrides -s)\n"
            "        -o <dir>       output directory used for -H\n"
            "        -r             recursively expand frameworks and fixed VM shared libraries\n"
            "        -s             sort classes and categories by name\n"
            "        -S             sort methods by name\n"
            "        -t             suppress header in output, for testing\n"
            "        --list-arches  list the arches in the file, then exit\n"
            "        --sdk-root     specify the SDK root path (full path, or 3.0, 3.1, 3.1.2, 3.1.3, 3.2, 4.0, 10.5, 10.6)\n"
            ,
            CLASS_DUMP_VERSION
       );
}

#define CD_OPT_ARCH 1
#define CD_OPT_LIST_ARCHES 2
#define CD_OPT_VERSION 3
#define CD_OPT_SDK_ROOT 4

struct sdk_alias {
    NSString *alias;
    NSString *path;
};

struct sdk_alias sdk_aliases[] = {
    { @"3.0", @"/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS3.0.sdk", },
    { @"3.1", @"/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS3.1.sdk", },
    { @"3.1.2", @"/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS3.1.2.sdk", },
    { @"3.1.3", @"/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS3.1.3.sdk", },
    { @"3.2", @"/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS3.2.sdk", },
    { @"4.0", @"/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS4.0.sdk", },
    { @"10.5", @"/Developer/SDKs/MacOSX10.5.sdk", },
    { @"10.6", @"/Developer/SDKs/MacOSX10.6.sdk", },
    { nil, nil, },
};

// Keyed by alias, value is full path.
NSDictionary *sdkAliases(void)
{
    static NSDictionary *aliases = nil;

    if (aliases == nil) {
        NSMutableDictionary *dict;
        struct sdk_alias *ptr = sdk_aliases;

        dict = [[NSMutableDictionary alloc] init];
        while (ptr->alias != nil) {
            [dict setObject:ptr->path forKey:ptr->alias];
            ptr++;
        }
        aliases = [dict copy];
        [dict release];
    }

    return aliases;
}

int main(int argc, char *argv[])
{
    NSAutoreleasePool *pool = [[NSAutoreleasePool alloc] init];
    CDClassDump *classDump;
    CDMultiFileVisitor *multiFileVisitor;
    BOOL shouldFind = NO;
    NSString *searchString = nil;
    BOOL shouldGenerateSeparateHeaders = NO;
    BOOL shouldListArches = NO;
    BOOL shouldPrintVersion = NO;
    CDArch targetArch;
    BOOL hasSpecifiedArch = NO;

    int ch;
    BOOL errorFlag = NO;

    struct option longopts[] = {
        { "show-ivar-offsets", no_argument, NULL, 'a' },
        { "show-imp-addr", no_argument, NULL, 'A' },
        { "match", required_argument, NULL, 'C' },
        { "find", required_argument, NULL, 'f' },
        { "generate-multiple-files", no_argument, NULL, 'H' },
        { "sort-by-inheritance", no_argument, NULL, 'I' },
        { "output-dir", required_argument, NULL, 'o' },
        { "recursive", no_argument, NULL, 'r' },
        { "sort", no_argument, NULL, 's' },
        { "sort-methods", no_argument, NULL, 'S' },
        { "arch", required_argument, NULL, CD_OPT_ARCH },
        { "list-arches", no_argument, NULL, CD_OPT_LIST_ARCHES },
        { "suppress-header", no_argument, NULL, 't' },
        { "version", no_argument, NULL, CD_OPT_VERSION },
        { "sdk-root", required_argument, NULL, CD_OPT_SDK_ROOT },
        { NULL, 0, NULL, 0 },
    };

    if (argc == 1) {
        print_usage();
        exit(0);
    }

    classDump = [[[CDClassDump alloc] init] autorelease];
    multiFileVisitor = [[[CDMultiFileVisitor alloc] init] autorelease];
    [multiFileVisitor setClassDump:classDump];

    while ( (ch = getopt_long(argc, argv, "aAC:f:HIo:rRsSt", longopts, NULL)) != -1) {
        switch (ch) {
          case CD_OPT_ARCH: {
              NSString *name;

              name = [NSString stringWithUTF8String:optarg];
              targetArch = CDArchFromName(name);
              if (targetArch.cputype != CPU_TYPE_ANY)
                  hasSpecifiedArch = YES;
              else {
                  fprintf(stderr, "Error: Unknown arch %s\n\n", optarg);
                  errorFlag = YES;
              }
              break;
          }

          case CD_OPT_LIST_ARCHES:
              shouldListArches = YES;
              break;

          case CD_OPT_VERSION:
              shouldPrintVersion = YES;
              break;

          case CD_OPT_SDK_ROOT: {
              NSString *root, *str;
              NSDictionary *aliases = sdkAliases();

              root = [NSString stringWithUTF8String:optarg];
              //NSLog(@"root: %@", root);
              //NSLog(@"aliases: %@", aliases);
              str = [aliases objectForKey:root];
              if (str == nil) {
                  classDump.sdkRoot = root;
              } else {
                  classDump.sdkRoot = str;
              }

              break;
          }

          case 'a':
              [classDump setShouldShowIvarOffsets:YES];
              break;

          case 'A':
              [classDump setShouldShowMethodAddresses:YES];
              break;

          case 'C': {
              NSString *errorMessage;

              if ([classDump setRegex:optarg errorMessage:&errorMessage] == NO) {
                  fprintf(stderr, "class-dump: Error with regex: '%s'\n\n", [errorMessage UTF8String]);
                  errorFlag = YES;
              }
              // Last one wins now.
              break;
          }

          case 'f': {
              shouldFind = YES;

              searchString = [NSString stringWithUTF8String:optarg];
              break;
          }

          case 'H':
              shouldGenerateSeparateHeaders = YES;
              break;

          case 'I':
              [classDump setShouldSortClassesByInheritance:YES];
              break;

          case 'o':
              [multiFileVisitor setOutputPath:[NSString stringWithUTF8String:optarg]];
              break;

          case 'r':
              [classDump setShouldProcessRecursively:YES];
              break;

          case 's':
              [classDump setShouldSortClasses:YES];
              break;

          case 'S':
              [classDump setShouldSortMethods:YES];
              break;

          case 't':
              [classDump setShouldShowHeader:NO];
              break;

          case '?':
          default:
              errorFlag = YES;
              break;
        }
    }

    if (errorFlag) {
        print_usage();
        exit(2);
    }

    if (shouldPrintVersion) {
        printf("class-dump %s compiled %s\n", CLASS_DUMP_VERSION, __DATE__ " " __TIME__);
        exit(0);
    }

    if (optind < argc) {
        NSString *arg, *executablePath;

        arg = [NSString stringWithFileSystemRepresentation:argv[optind]];
        executablePath = [arg executablePathForFilename];
        if (shouldListArches) {
            if (executablePath == nil) {
                printf("none\n");
            } else {
                CDSearchPathState *searchPathState;
                NSData *data;
                id macho;

                searchPathState = [[[CDSearchPathState alloc] init] autorelease];
                searchPathState.executablePath = executablePath;
                data = [[NSData alloc] initWithContentsOfMappedFile:executablePath];
                macho = [CDFile fileWithData:data filename:executablePath searchPathState:searchPathState];
                if (macho == nil) {
                    printf("none\n");
                } else {
                    if ([macho isKindOfClass:[CDMachOFile class]]) {
                        printf("%s\n", [[macho archName] UTF8String]);
                    } else if ([macho isKindOfClass:[CDFatFile class]]) {
                        printf("%s\n", [[[macho archNames] componentsJoinedByString:@" "] UTF8String]);
                    }
                }
                [data release];
            }
        } else {
            CDFile *file;
            NSData *data;

            if (executablePath == nil) {
                fprintf(stderr, "class-dump: Input file (%s) doesn't contain an executable.\n", [arg fileSystemRepresentation]);
                exit(1);
            }

            data = [[NSData alloc] initWithContentsOfMappedFile:executablePath];
            if (data == nil) {
                NSFileManager *defaultManager = [NSFileManager defaultManager];

                if ([defaultManager fileExistsAtPath:executablePath]) {
                    fprintf(stderr, "class-dump: Input file (%s) is not readable (check read rights).\n", [executablePath UTF8String]);
                } else {
                    fprintf(stderr, "class-dump: Input file (%s) does not exist.\n", [executablePath UTF8String]);
                }

                exit(1);
            }

            classDump.searchPathState.executablePath = [executablePath stringByDeletingLastPathComponent];
            file = [CDFile fileWithData:data filename:executablePath searchPathState:classDump.searchPathState];
            if (file == nil) {
                fprintf(stderr, "class-dump: Input file (%s) is neither a Mach-O file nor a fat archive.\n", [executablePath UTF8String]);
                [data release];
                exit(1);
            }
#if 0
            {
                CDFatFile *fat = file;
                NSArray *a1;
                NSUInteger count, index;

                a1 = [fat arches];
                count = [a1 count];
                for (index = 0; index < count; index++)
                    [[[a1 objectAtIndex:index] machOData] writeToFile:[NSString stringWithFormat:@"/tmp/arch-%u", index] atomically:NO];

                exit(99);
            }
#endif
            if (hasSpecifiedArch == NO) {
                if ([file bestMatchForLocalArch:&targetArch] == NO) {
                    fprintf(stderr, "Error: Couldn't get local architecture\n");
                    exit(1);
                }
                //NSLog(@"No arch specified, best match for local arch is: (%08x, %08x)", targetArch.cputype, targetArch.cpusubtype);
            } else {
                //NSLog(@"chosen arch is: (%08x, %08x)", targetArch.cputype, targetArch.cpusubtype);
            }

#ifndef __LP64__
            if (CDArchUses64BitABI(targetArch)) {
                fprintf(stderr, "Error: Can't dump 64-bit files with 32-bit version of class-dump\n");
                exit(1);
            }
#endif

            [classDump setTargetArch:targetArch];
            classDump.searchPathState.executablePath = [executablePath stringByDeletingLastPathComponent];

            if ([classDump loadFile:file]) {
#if 0
                [classDump showHeader];
                [classDump showLoadCommands];
                exit(5);
#endif

                [classDump processObjectiveCData];
                [classDump registerTypes];

                if (shouldFind) {
                    CDFindMethodVisitor *visitor;

                    visitor = [[CDFindMethodVisitor alloc] init];
                    [visitor setClassDump:classDump];
                    [visitor setFindString:searchString];
                    [classDump recursivelyVisit:visitor];
                    [visitor release];
                } else if (shouldGenerateSeparateHeaders) {
                    [classDump recursivelyVisit:multiFileVisitor];
                } else {
                    CDClassDumpVisitor *visitor;

                    visitor = [[CDClassDumpVisitor alloc] init];
                    [visitor setClassDump:classDump];
                    [classDump recursivelyVisit:visitor];
                    [visitor release];
                }
            }

            [data release];
        }
    }

    [pool release];

    return 0;
}
